Clarify pendulum use in timezone cases

potiuk opened the pull request:
It is important to use Pendulum in case timezone is used - because
there are a number of limitations coming from using stdlib
timezone implementation.

However our documentation was not very clear about it, especially
some examples shown using standard datetime in DAGs which could
mislead our users to continue using datetime if they use timezone.

This PR clarifies and stresses the use of pendulum is necessary
when timezone is used. Also it points to the documentation
in case serialization throws error about not using Pendulum
so that the users can learn about the reasoning.

See also #20070

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

kaxil commented on docs/apache-airflow/faq.rst:
```diff
@@ -243,6 +245,14 @@ This allows for a backfill on tasks that have ``depends_on_past=True`` to
+Using timezones
+---------------
+
+Creating a time zone aware time is quite simple. Just make sure to supply a time zone aware dates
```
```suggestion
Creating a time zone aware datetime (e.g. DAG's `start_date`) is quite simple. Just make sure to supply a time zone aware dates
```

kaxil commented on docs/apache-airflow/tutorial.rst:
```diff
@@ -230,6 +230,14 @@ Note that when executing your script, Airflow will raise exceptions when
+Using timezones
+---------------
+
+Creating a time zone aware time is quite simple. Just make sure to supply a time zone aware dates
```
```suggestion
Creating a time zone aware DAG is quite simple. Just make sure to supply a time zone aware dates
```

kaxil reviewed (approved):
nits - lgtm

jedcunningham commented on docs/apache-airflow/timezone.rst:
```diff
@@ -120,11 +121,18 @@ it is therefore important to make sure this setting is equal on all Airflow node
+have limitations and we deliberately disallow using them in DAGs.
+
+, as this will throw serialization
+errors.
```
```suggestion
```

Leftovers?

potiuk replied:
yep.

jedcunningham commented on docs/apache-airflow/faq.rst:
```diff
@@ -243,6 +245,14 @@ This allows for a backfill on tasks that have ``depends_on_past=True`` to
+Using timezones
+---------------
+
+Creating a time zone aware datetime (e.g. DAG's `start_date`) is quite simple. Just make sure to supply a time zone aware dates
```
```suggestion
Creating a time zone aware datetime (e.g. DAG's ``start_date``) is quite simple. Just make sure to supply a time zone aware dates
```

josh-fell commented on docs/apache-airflow/dag-run.rst:
```diff
@@ -111,7 +111,9 @@ in the configuration file. When turned off, the scheduler creates a DAG run only
-    from datetime import datetime, timedelta
+
+    # Note that if you plan to use timezones you should use pendulum
+    from pendulum import datetime, timedelta
```
Do we want to also push use of `pendulum.duration()` (the `datetime.timedelta()` equivalent) too?

josh-fell replied:
Actually that would require some other code changes I think. Just need to update these docs to still use `datetime.timedelta` since pendulum doesn't have `timedelta`.

potiuk replied:
Ah good point indeed let me try that one and see it.

potiuk replied:
I don't think in this case we need ot to do it. The thing here is to make the "datetime" pendulum, so that whenever you want to add timezone (for example because you realise you need it after the fact), you are not tempted to  use timezone from stdlib - because it causes problems. For example however it would likely be good to separate it visibly at least

potiuk replied:
Well. It is still a mess. Pendulum choosing to implement drop-in replacement for PARTS (?!*&%$) of datetime package  was the next biggest mistake after Python2 encoding choices.

potiuk replied:
Actually I think much better solution will be to add tz to all the examples. I think this is the BEST way to show people best practices.

potiuk committed:
Clarify pendulum use in timezone cases

It is important to use Pendulum in case timezone is used - because
there are a number of limitations coming from using stdlib
timezone implementation.

However our documentation was not very clear about it, especially
some examples shown using standard datetime in DAGs which could
mislead our users to continue using datetime if they use timezone.

This PR clarifies and stresses the use of pendulum is necessary
when timezone is used. Also it points to the documentation
in case serialization throws error about not using Pendulum
so that the users can learn about the reasoning.

This is the first part of the change - the follow up will be
changing all provider examples to also use timezone and
pendulum explicitely.

See also #20070

jedcunningham commented on UPDATING.md:
```diff
@@ -245,7 +245,7 @@ Similarly, `DAG.concurrency` has been renamed to `DAG.max_active_tasks`.
 dag = DAG(
     dag_id="example_dag",
-    start_date=datetime(2021, 1, 1),
+    start_date=pendulum.datetime(2021, 1, 1, tz="US/Eastern"),
```
```suggestion
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
```

Might just be me, but I'd rather see UTC here.

potiuk replied:
I am fine with that. Let me global replace it

potiuk replied:
Pushed

jedcunningham commented on airflow/example_dags/example_bash_operator.py:
```diff
@@ -27,9 +29,9 @@
     dag_id='example_bash_operator',
     schedule_interval='0 0 * * *',
-    start_date=datetime(2021, 1, 1),
+    start_date=pendulum.datetime(2021, 1, 1),
```
```suggestion
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
```

Was this just missed?

jedcunningham commented on airflow/example_dags/example_branch_datetime_operator.py:
```diff
@@ -20,15 +20,15 @@
 dag = DAG(
     dag_id="example_branch_datetime_operator",
-    start_date=datetime.datetime(2021, 1, 1),
+    start_date=pendulum.datetime(2021, 1, 1),
```
```suggestion
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
```

and here?

potiuk committed:
Update airflow/example_dags/example_bash_operator.py

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

potiuk committed:
Update airflow/example_dags/example_branch_datetime_operator.py

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>
